contracts/PriceOracle/PriceOracle.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.7;
+
+import { Ownable } from "../Ownable.sol";
+import { AggregatorV3Interface } from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+import { FullMath } from "../Uniswap/UniswapLib.sol";
+
+contract PriceOracle is Ownable {
+
+    /// @dev Configuration used to return the USD price for the associated cToken asset and base unit needed for formatting
+    /// There should be 1 TokenConfig object for each supported asset, passed in the constructor.
+    struct TokenConfig {
+        // The address of the Compound Token
+        address cToken;
+        // The number of smallest units of measurement in a single whole unit. (e.g. 1e18 for ETH)
+        uint256 baseUnit;
+        // Address of the feed used to retrieve the asset's price
+        address priceFeed;
+    }
+
+    /// @dev Mapping of cToken address to TokenConfig used to maintain the supported assets
+    mapping (address => TokenConfig) tokenConfigs;
+
+    /// @notice The event emitted when a new asset is added to the mapping
+    event PriceOracleAssetAdded(address indexed cToken, uint256 baseUnit, address priceFeed);
+
+    /// @notice The event emitted when the price feed for an existing config is updated
+    event PriceOracleAssetPriceFeedUpdated(address indexed cToken, address oldPriceFeed, address newPriceFeed);
+
+    /// @notice The event emitted when an asset is removed to the mapping
+    event PriceOracleAssetRemoved(address indexed cToken, uint256 baseUnit, address priceFeed);
+
+    /**
+     * @notice Construct a Price Oracle contract for a set of token configurations
+     * @param configs The token configurations that define which price feed and base unit to use for each asset
+     */
+    constructor(TokenConfig[] memory configs) {
+        // Populate token config mapping 
+        for (uint i = 0; i < configs.length; i++) {
+            TokenConfig memory config = configs[i];
+            validateTokenConfig(config);
+            // Check if duplicate configs were submitted for the same cToken
+            require(tokenConfigs[config.cToken].cToken == address(0), "Duplicate config for cToken found");
+            tokenConfigs[config.cToken] = config;
+        }
+    }
+
+    /**
+     * @notice Get the underlying price of a cToken, in the format expected by the Comptroller.
+     * @dev Implements the PriceOracle interface for Compound v2.
+     * @param cToken The cToken address for price retrieval
+     * @return Price denominated in USD for the given cToken address, in the format expected by the Comptroller.
+     */
+    function getUnderlyingPrice(address cToken)
+        external
+        view
+        returns (uint256)
+    {
+        TokenConfig memory config = tokenConfigs[cToken];
+        require(config.cToken != address(0), "Config not found for cToken");
```

Comment:
Consider using custom revert statements instead of `require`.  

https://kristaps.me/blog/solidity-custom-error/#:~:text=Using%20Custom%20Error,and%20addressing%20issues%20more%20efficiently.